### ISSUE TITLE
Add additional address information for USA

### ISSUE THREAD

**Issue by Doug-Murphy** [nbuilder/nbuilder#104]
Support should be added for additional address information for USA beyond just the state. This includes address line 1, city and postal (ZIP) code.

**Comment by PureKrome** [comment]
![](https://i.ytimg.com/vi/KtJJC0nWs9s/maxresdefault.jpg)

that said, this would be a great "Up for Grabs" issue for those who can submit a PR.

**Comment by Doug-Murphy** [comment]
@PureKrome I already have. :)

**Comment by PureKrome** [comment]
Yep 😊 - just edited my reply. 😊

**Comment by crmckenzie** [comment]
I'm conflicted on this one. It's the sort of feature that's potentially never done since there will always be a random city, state, county, province, district, locale, etc. that gets missed. I'm thinking NBuilder may have taken a wrong approach here. Perhaps it would be better to allow property assignment from a provided set of values.

I'm pseudocoding here, but something like:

```csharp
var cities = new ValueSet<String>("Chicago", "New York", "Seattle", "Miami", "Los Angeles");
Builder<Product>
	.CreateListOfSize(10)
	.All()
	.Have(x => x.City = cities.Next())
	.Build();
```

Thoughts?

**Comment by Doug-Murphy** [comment]
@crmckenzie Yeah I think that's a fine idea honestly. I see this functionality co-existing with your proposal above as there is already the `GetRandom` static class. There are two different usages that I see being common and valid. Some people may use this package just for generating random data explicitly using the `GetRandom` static class into existing objects. Others might use the more NBuilder-esque approach as you psuedocoded above where it generates the full object. I think both are great and both should be well supported.

**Comment by PureKrome** [comment]
I prefer the property assignment method. It always vexxes Non-USA people (which are a lot, for-the-record) that the data is only USA related. 

Keep this data out of NBuilder and make it _really easy_ to use <your own data source>.

For example, I just did a 1 sec google search and found a [free list of USA cities/~postcodes~zipcodes/etc](https://simplemaps.com/data/us-cities). If i manually download this file, write a line or two of code to read this data into a list, then provide that list to Nbuider (`.Have(x => x.City = cities.Next())`) then we're good to go!

**Comment by crmckenzie** [comment]
Well, my thought was to introduce the property assignment method and mark the hard-coded lists as obsolete. That preserves the existing behavior for awhile while people get used to the newer, better method. `ValueSet` could also offer a `Random()` function as opposed to `Next()` just to keep things exciting and user-friendly.

> It always vexxes Non-USA people (which are a lot, for-the-record)

@PureKrome do you have any data to support this assertion? lolz

**Comment by PureKrome** [comment]
> Well, my thought was to introduce the property assignment method and mark the hard-coded lists as obsolete.

great! that's consistent with how MS does it (e.g. [the good ole Binaryformatter Serializer](https://github.com/dotnet/designs/pull/141)). In a future major semver, it can then get removed.

> do you have any data to support this assertion? lolz

:) i'll check with my American friends for some data.

oh

wait ....

**Comment by crmckenzie** [comment]
```csharp
        [Fact]
        public void Sample_UsingInABuilderContext()
        {
            var names = new ValueSet<string>("fredbob", "foobar", "pat boone", "debbie boone");

            var demos = new Builder()
                .CreateListOfSize<Demo>(5)
                .All()
                .With(row => row.Name = names.Next())
                .Build()
                ;

            demos[0].Name.ShouldBe("fredbob");
            demos[1].Name.ShouldBe("foobar");
            demos[4].Name.ShouldBe("fredbob");

        }
```
I just implemented this and it works nicely.

**Comment by crmckenzie** [comment]
```csharp
        [Fact]
        public void Random_ValuesAreGeneratedSomewhatRandomly()
        {
            var values = new string[] { "fredbob", "foobar", "pat boone", "debbie boone" };

            var results = new List<string>();
            for (var i = 0; i < 100; i++)
            {
                // The idea here is to create a new set every time with
                // enough delay so that seeding the random number generator
                // yields a different random sequence. By creatinga new set
                // every time, collecting the results, and asserting we have
                // a complete list of values, we confirm that we are indeed
                // getting all of the values even in the first call to Random()
                // on the set. This proves that we're not just reimplementing
                // Next()
                //
                System.Threading.Thread.Sleep(10);
                var set = new ValueSet<string>(values);
                results.Add(set.Random());
            }

            var unique = results.Distinct().OrderBy(value => value).ToHashSet<string>();
            var compare = values.OrderBy(value => value).ToHashSet<string>();

            unique.ShouldBe(compare);
        }

        [Fact]
        public void Random_AllValuesAreRepresentedInASufficientlyLargeDistribution()
        {
            var set = new ValueSet<string>("fredbob", "foobar", "pat boone", "debbie boone");
            
            var results = new List<string>();
            for (var i = 0; i< 100; i++)
            {
                results.Add(set.Random());
            }

            var unique = results.Distinct().OrderBy(value => value).ToHashSet<string>();
            var compare = set.OrderBy(value => value).ToHashSet<string>();

            unique.ShouldBe(compare);
       }
```

**Comment by Doug-Murphy** [comment]
I thought I understood the new proposal at first, and now I think I don't. :D @crmckenzie Are you proposing that the consumer of NBuilder will be responsible for providing their own list for generating the random pieces like first/last names, cities, etc? I can kind of agree with this because it adds more flexibility, but I also find it an inconvenience. Personally speaking, I found that NBuilder having a built-in list already for me to be a huge benefit! I'm not sure which case would be more likely when we're talking about if the consumer would want to control the random data. What I mean by that is "Does it matter what the city winds up being?" Probably not. We just want something that is syntactically speaking, a valid city (for simplicity, let's just say that's any alphabetic string).

I think having a defaulted list inside of NBuilder is very useful, but providing the flexibility for the consumer to override those values is also very useful. Basically this makes it easy and useful no matter the scenario the consumer is in.

**Comment by crmckenzie** [comment]
> Are you proposing that the consumer of NBuilder will be responsible for providing their own list for generating the random pieces like first/last names, cities, etc?

I'm proposing preserving what's there, but moving it out of `GetRandom` while providing the ability for end-users to customize their needs through `ValueSet`.

**Comment by Doug-Murphy** [comment]
Gotcha. In that case, you have my support! :D I'm confused about `ValueSet` though. I haven't used it before, but it seems like `HashSet` may be a better data structure based on what I perceive the usecase to be.

**Comment by crmckenzie** [comment]
I'll try to explain ValueSet.

Consider an enum.

```csharp
public enum EmployeeType {
    Salaried,
    Hourly
}
```

When assigning data to NBuilder collections, I want it to iterate over the enum variants so that they are distributed through my test data.

```csharp
employees[0].type.ShouldBe(EmployeeType.Salaried);
employees[1].type.ShouldBe(EmployeeType.Hourly);
employees[2].type.ShouldBe(EmployeeType.Salaried);
employees[3].type.ShouldBe(EmployeeType.Hourly);
```

What happens here is that when we reach the end the list of variants, NBuilder starts over from the beginning of the list. 

What happens with `GetRandom` is different--it always chooses a random value from some list. 

What we have between these two constructs is two different concerns: the range of values to select from for a particular property, and the selection strategy.

`ValueSet<T>` captures a range of particular values and provides two different selection strategies via the `Next()` and `Random()` functions. When using `Next()`, the selection strategy is _sequential_, starting over from the beginning when the set is exhausted. When using `Random()` the value assigned will always be random. I've personally found use-cases for both in my own code to this seems like a win to me.

Thoughts?

**Comment by Doug-Murphy** [comment]
Well it helps that now I see `ValueSet` is a custom data structure. :D Haha. Yes this all makes perfect sense now.

So in the interest of moving forward, what if we merge the associated PR and close out this issue? Then look at the `ValueSet` work again since it seems to be a fundamental redesign. These two things seem disjointed enough such that one should not block the other.
